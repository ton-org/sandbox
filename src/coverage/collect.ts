import { Address, beginCell, Cell } from '@ton/core';
import { runtime, text, trace } from 'ton-assembly';

import { BlockchainTransaction } from '../blockchain/Blockchain';
import { buildLineInfo, CoverageData } from './data';

export function collectAsmCoverage(cell: Cell, logs: string): CoverageData {
    const [cleanCell, mapping] = recompileCell(cell);
    const info = trace.createMappingInfo(mapping);

    const traceInfos = trace.createTraceInfoPerTransaction(logs, info, undefined);
    const assembly = text.print(runtime.decompileCell(cleanCell));
    const combinedTrace = { steps: traceInfos.flatMap((trace) => trace.steps) };
    const combinedLines = buildLineInfo(combinedTrace, assembly);
    return {
        code: cell,
        lines: combinedLines,
    };
}

function recompileCell(cell: Cell): [Cell, runtime.Mapping] {
    const instructionsWithoutPositions = runtime.decompileCell(cell);
    const assemblyForPositions = text.print(instructionsWithoutPositions);

    const parseResult = text.parse('out.tasm', assemblyForPositions);
    if (parseResult.$ === 'ParseFailure') {
        throw new Error('Cannot parse resulting text Assembly');
    }

    return runtime.compileCellWithMapping(parseResult.instructions);
}

export function collectTxsCoverage(
    code: Cell,
    address: Address | undefined,
    transactions: readonly BlockchainTransaction[],
): CoverageData[] {
    const results: CoverageData[] = [];

    for (const transaction of transactions) {
        const txAddress = bigintToAddress(transaction.address);
        if (address !== undefined && txAddress?.toString() !== address.toString()) {
            // other contract transaction, skip
            continue;
        }

        results.push(collectAsmCoverage(code, transaction.vmLogs));
    }

    return results;
}

function bigintToAddress(addr: bigint | undefined): Address | undefined {
    if (addr === undefined) return undefined;

    try {
        const slice = beginCell().storeUint(4, 3).storeUint(0, 8).storeUint(addr, 256).asSlice();
        return slice.loadAddress();
    } catch {
        return undefined;
    }
}
