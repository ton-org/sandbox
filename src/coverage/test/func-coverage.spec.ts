import {runtime} from "ton-assembly";
import {generateTextReport, generateHtmlReport} from "../";
import {mkdirSync, writeFileSync, existsSync} from "node:fs";
import {executeInstructions} from "./execute";
import {collectAsmCoverage} from "../collect";
import {compileFunc} from "@ton-community/func-js";
import {Cell} from "@ton/core";
import {decompileCell} from "ton-assembly/dist/runtime";

describe("func asm coverage", () => {
    const test =
        (code: string, id: number = 0) =>
            async () => {
                const name = expect.getState().currentTestName;

                const funcCompiled = await compile(code);
                const funcInstructions = decompileCell(funcCompiled);

                const cell = runtime.compileCell(funcInstructions);
                const [_, logs] = await executeInstructions(funcInstructions, id);
                const coverage = collectAsmCoverage(cell, logs);

                const report = generateTextReport(coverage);
                expect(report).toMatchSnapshot();

                const outDirname = `${__dirname}/output`;
                if (!existsSync(outDirname)) {
                    mkdirSync(outDirname);
                }

                const htmlReport = generateHtmlReport(coverage);
                writeFileSync(`${__dirname}/output/func-${name}.html`, htmlReport);
            };

    it(
        "simple if",
        test(
            `
                global int foo;
            
                () recv_internal() {
                    foo = 10;
                    if (foo > 10) {
                        foo = 20;
                        return ();
                    }
                    
                    foo = 15;
                    return ();
                }
            `,
        ),
    );
});

const compile = async (code: string): Promise<Cell> => {
    const res = await compileFunc({
        sources: [
            {
                content: code,
                filename: "source",
            },
        ],
    });
    if (res.status === "error") {
        throw new Error(`cannot compile FunC: ${res.message}`);
    }

    return Cell.fromBase64(res.codeBoc);
};
