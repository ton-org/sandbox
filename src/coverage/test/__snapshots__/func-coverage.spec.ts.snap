// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`func asm coverage simple if 1`] = `
"Coverage Summary:
Lines: 11/14 (78.57%)
Total Gas: 351
Total Hits: 11

Instruction Stats:
  DICTIGETJMPZ   | 126 gas | 1 hits |  126 avg gas |  35.90%
  DICTPUSHCONST  |  34 gas | 1 hits |   34 avg gas |   9.69%
  SETGLOB        |  33 gas | 2 hits | 16.5 avg gas |   9.40%
  SETCP          |  26 gas | 1 hits |   26 avg gas |   7.41%
  GETGLOB        |  26 gas | 1 hits |   26 avg gas |   7.41%
  GTINT          |  26 gas | 1 hits |   26 avg gas |   7.41%
  PUSHINT_8      |  26 gas | 1 hits |   26 avg gas |   7.41%
  PUSHINT_4      |  18 gas | 1 hits |   18 avg gas |   5.13%
  PUSHCONT_SHORT |  18 gas | 1 hits |   18 avg gas |   5.13%
  IFJMP          |  18 gas | 1 hits |   18 avg gas |   5.13%

Annotated Code:
 1 ✓ | SETCP 0                                  | gas:26    | hits:1
 2 ✓ | DICTPUSHCONST 19 [                       | gas:34    | hits:1
 3   |     0 => {                               |           |
 4 ✓ |         PUSHINT_4 10                     | gas:18    | hits:1
 5 ✓ |         SETGLOB 1                        | gas:28    | hits:1
 6 ✓ |         GETGLOB 1                        | gas:26    | hits:1
 7 ✓ |         GTINT 10                         | gas:26    | hits:1
 8 ✓ |         PUSHCONT_SHORT {                 | gas:18    | hits:1
 9 ✗ |             PUSHINT_8 20                 |           |
10 ✗ |             SETGLOB 1                    |           |
11   |         }                                |           |
12 ✓ |         IFJMP                            | gas:18    | hits:1
13 ✓ |         PUSHINT_8 15                     | gas:26    | hits:1
14 ✓ |         SETGLOB 1                        | gas:5     | hits:1
15   |     }                                    |           |
16   | ]                                        |           |
17 ✓ | DICTIGETJMPZ                             | gas:126   | hits:1
18 ✗ | THROWARG 11                              |           |
19   |                                          |           |"
`;

exports[`func asm coverage simple if 2 1`] = `
"Coverage Summary:
Lines: 11/14 (78.57%)
Total Gas: 359
Total Hits: 11

Instruction Stats:
  DICTIGETJMPZ   | 126 gas | 1 hits |  126 avg gas |  35.10%
  PUSHINT_8      |  52 gas | 2 hits |   26 avg gas |  14.48%
  DICTPUSHCONST  |  34 gas | 1 hits |   34 avg gas |   9.47%
  SETGLOB        |  33 gas | 2 hits | 16.5 avg gas |   9.19%
  SETCP          |  26 gas | 1 hits |   26 avg gas |   7.24%
  GETGLOB        |  26 gas | 1 hits |   26 avg gas |   7.24%
  GTINT          |  26 gas | 1 hits |   26 avg gas |   7.24%
  PUSHCONT_SHORT |  18 gas | 1 hits |   18 avg gas |   5.01%
  IFJMP          |  18 gas | 1 hits |   18 avg gas |   5.01%

Annotated Code:
 1 ✓ | SETCP 0                                  | gas:26    | hits:1
 2 ✓ | DICTPUSHCONST 19 [                       | gas:34    | hits:1
 3   |     0 => {                               |           |
 4 ✓ |         PUSHINT_8 15                     | gas:26    | hits:1
 5 ✓ |         SETGLOB 1                        | gas:28    | hits:1
 6 ✓ |         GETGLOB 1                        | gas:26    | hits:1
 7 ✓ |         GTINT 10                         | gas:26    | hits:1
 8 ✓ |         PUSHCONT_SHORT {                 | gas:18    | hits:1
 9 ✓ |             PUSHINT_8 20                 | gas:26    | hits:1
10 ✓ |             SETGLOB 1                    | gas:5     | hits:1
11   |         }                                |           |
12 ✓ |         IFJMP                            | gas:18    | hits:1
13 ✗ |         PUSHINT_8 15                     |           |
14 ✗ |         SETGLOB 1                        |           |
15   |     }                                    |           |
16   | ]                                        |           |
17 ✓ | DICTIGETJMPZ                             | gas:126   | hits:1
18 ✗ | THROWARG 11                              |           |
19   |                                          |           |"
`;
